priam/src/main/java/com/netflix/priam/backupv2/BackupVerificationTask.java

```diff
@@ -88,22 +89,32 @@ public void execute() throws Exception {
                                         backupRestoreConfig.getBackupVerificationSLOInHours(),
                                         ChronoUnit.HOURS),
                         DateUtil.getInstant());
-        Optional<BackupVerificationResult> verificationResult =
-                backupVerification.verifyBackup(
+        List<BackupVerificationResult> verificationResults =
+                backupVerification.verifyAllBackups(
                         BackupVersion.SNAPSHOT_META_SERVICE, false, dateRange);
-        if (!verificationResult.isPresent() || !verificationResult.get().valid) {
+        if (verificationResults.isEmpty()
+                || verificationResults
+                        .stream()
+                        .filter(r -> r.valid)
+                        .collect(Collectors.toList())
+                        .isEmpty()) {
             logger.error(
                     "Not able to find any snapshot which is valid in our SLO window: {} hours",
                     backupRestoreConfig.getBackupVerificationSLOInHours());
             backupMetrics.incrementBackupVerificationFailure();
         } else {
             // verification result is available and is valid.
             // send notification that backup is uploaded.
-            logger.info(
-                    "Sending {} message for backup: {}",
-                    AbstractBackupPath.BackupFileType.SNAPSHOT_VERIFIED,
-                    verificationResult.get().snapshotInstant);
-            backupNotificationMgr.notify(verificationResult.get());
+            verificationResults
+                    .stream()
+                    .forEach(
+                            r -> {
+                                logger.info(
+                                        "Sending {} message for backup: {}",
+                                        AbstractBackupPath.BackupFileType.SNAPSHOT_VERIFIED,
+                                        r.snapshotInstant);
+                                backupNotificationMgr.notify(r);
```

Comment:
As per your method of verifyAllBackups we add all the backups to the list (even ones which were verified before). This will basically make the logic as - "send verifications for all the backups in the SLO window every time we run this job which is every hour". Thus we will end up sending 48 notifications every hour for every node. It does get the job done but we are sending too many notifications which dont need to be sent. 
We should send notifications for all the "newly" verified snapshots and the latest snapshot ONLY. Most of the times we will send notification for "latest" snapshot with sometimes sending past message.

Reply:
Makes sense. Let me see how I can optimize the number of verifications and thereby notifications.